core/src/main/java/com/amplifyframework/auth/result/AuthSessionResult.java

```diff
@@ -0,0 +1,133 @@
+     * @return A new result object for the successfully retrieved value.
+     */
+    public static <T> AuthSessionResult<T> success(T value) {
+        return new AuthSessionResult<>(value, null, Type.SUCCESS);
+    }
+
+    /**
+     * Create a result object for a failed value fetch operation.
+     * @param error The error describing what went wrong while trying to fetch the value.
+     * @param <T> The type of value attempting to be retrieved.
+     * @return A new result object for the unsuccessfully retrieved value.
+     */
+    public static <T> AuthSessionResult<T> failure(AuthException error) {
+        return new AuthSessionResult<>(null, error, Type.FAILURE);
+    }
+
+    /**
+     * Returns the value which was successfully retrieved.
+     * @return the value which was successfully retrieved.
+     */
+    @Nullable
+    public T getValue() {
+        return value;
+    }
+
+    /**
+     * Returns the error describing what went wrong while attempting to retrieve the value.
+     * @return the error describing what went wrong while attempting to retrieve the value.
+     */
+    @Nullable
+    public AuthException getError() {
+        return error;
+    }
+
+    /**
+     * Returns the type of result this was - whether it was successful or a failure. Useful for switch statements.
+     * @return the type of result this was - whether it was successful or a failure.
+     */
+    @NonNull
+    public Type getType() {
+        return type;
+    }
+
+    @Override
+    public int hashCode() {
+        return ObjectsCompat.hash(
+                getValue(),
+                getError(),
+                getType()
+        );
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        if (this == obj) {
+            return true;
+        } else if (obj == null || !(obj instanceof AuthSessionResult<?>)) {
+            return false;
+        } else {
+            AuthSessionResult<?> authSessionResult = (AuthSessionResult<?>) obj;
+            return ObjectsCompat.equals(getValue(), authSessionResult.getValue()) &&
+                    ObjectsCompat.equals(getError(), authSessionResult.getError()) &&
+                    ObjectsCompat.equals(getType(), authSessionResult.getType());
+        }
+    }
+
+    @Override
+    public String toString() {
+        return "AuthSessionResult{" +
+                "value=" + getValue() +
+                ", error=" + getError() +
+                ", type=" + getType() +
+                '}';
+    }
+
+    /**
+     * The type of result this was - whether it was successful or a failure.
+     */
+    public enum Type {
+        /**
+         * Fetch value operation was successful.
+         * You can retrieve the result with {@link #getValue()}.
+         */
+        SUCCESS,
+        /**
+         * Fetch value operation failed.
+         * You can retrieve the exception explaining what went wrong with {@link #getError()}.
+         */
+        FAILURE
+    }
```

Comment:
is it always either-or situation? Is it possible to have partial result, where both result & non-blocking error are returned? 

Also, can't this be represented with a simple boolean `isSuccessful` or something? Otherwise, maybe we can rename to `Status` or `State` instead? `AuthSessionResult#getType()` seems very vague, and gives off feeling that authentication is categorized into multiple types of operation.

Reply:
Yeah it is always either or - basically it's designed the way it is to mirror the Swift result object: https://www.swiftbysundell.com/articles/the-power-of-result-types-in-swift/

Also mirrors the design Jameson did for an object like this on Android with a getType method.

Reply:
I guess using a specific enum to represent the state over boolean is fine, and it's not unprecedented as you mention. But I think we should still strive for a clearer name especially for customer-exposed data structures like this. Also, `Status` is what I'm leaning towards as the name, similarly to [this](https://github.com/aws-amplify/amplify-android/blob/master/core/src/main/java/com/amplifyframework/core/InitializationStatus.java).

Reply:
Discussed offline that the idea of type in the context of request makes sense since there are static success/failure type results vs. state which reflects it could change and that this was the naming decision Jameson went with in mocking out the class: https://gist.github.com/jamesonwilliams/b15b8888c2e759d575c59fb1567ea32f